fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -2432,65 +2456,87 @@ void MultiVersionApi::updateSupportedVersions() {
 	}
 }
 
-ThreadFuture<Void> MultiVersionApi::updateClusterSharedStateMap(std::string clusterFilePath,
-                                                                ProtocolVersion dbProtocolVersion,
-                                                                Reference<IDatabase> db) {
-	MutexHolder holder(lock);
-	if (clusterSharedStateMap.find(clusterFilePath) == clusterSharedStateMap.end()) {
+// Must be called from the main thread
+ACTOR Future<Void> updateClusterSharedStateMapImpl(MultiVersionApi* self,
+                                                   ClusterConnectionRecord connectionRecord,
+                                                   ProtocolVersion dbProtocolVersion,
+                                                   Reference<IDatabase> db) {
+	// The cluster ID will be the connection record string (either a filename or the connection string itself)
+	// in API versions before we could read the cluster ID.
+	state std::string clusterId = connectionRecord.toString();
+	if (MultiVersionApi::apiVersionAtLeast(720)) {
+		state Reference<ITransaction> tr = db->createTransaction();
+		loop {
+			try {
+				state ThreadFuture<Optional<Value>> clusterIdFuture = tr->get("\xff\xff/cluster_id"_sr);
```

Comment:
Would it be more efficient if we call the getClusterId() function you added to the native API? Of course we need then to expose it over the internal API.

Reply:
I did it this way based on previous team discussion that this kind of data would be preferable to expose this way rather than adding new C APIs for each bit of data we wanted to expose. It's similar to the functionality to read the connection string or cluster file path, etc.

Reply:
I like the idea of special keys and I agree to that we should use this approach as long as it does not affect performance significantly or is not on a performance-critical path. I am not quite sure what overheads this approach is introducing compared to simply retrieving the clusterID from the ClientDBInfo. If it requires at least one additional network roundtrip, I am afraid it will affect database connection time significantly.